### ISSUE TITLE
Build failed: Java 11 perhaps

### ISSUE THREAD

**Issue by binkley** [charleskorn/batect#41]
```
batect.integrationtests.DockerClientIntegrationTest > describe a Docker client FAILED
    org.mockito.exceptions.base.MockitoException: 
    Mockito cannot mock this class: class batect.logging.Logger.
    Can not mock final classes with the following settings :
     - explicit serialization (e.g. withSettings().serializable())
     - extra interfaces (e.g. withSettings().extraInterfaces(...))

    You are seeing this disclaimer because Mockito is configured to create inlined mocks.
    You can learn about inline mocks and their limitations under item #39 of the Mockito class javadoc.

    Underlying exception : org.mockito.exceptions.base.MockitoException: Could not modify all classes [class java.lang.Object, class batect.logging.Logger]
        at batect.integrationtests.DockerClientIntegrationTest$1$1.invoke(DockerClientIntegrationTest.kt:334)
        at batect.integrationtests.DockerClientIntegrationTest$1$1.invoke(DockerClientIntegrationTest.kt:68)
        at org.jetbrains.spek.engine.SpekTestEngine$Collector.group(SpekTestEngine.kt:152)
        at org.jetbrains.spek.api.dsl.SpecBody$DefaultImpls.group$default(SpecBody.kt:13)
        at org.jetbrains.spek.api.dsl.StandardKt.describe(Standard.kt:10)
        at batect.integrationtests.DockerClientIntegrationTest$1.invoke(DockerClientIntegrationTest.kt:69)
        at batect.integrationtests.DockerClientIntegrationTest$1.invoke(DockerClientIntegrationTest.kt:68)

        Caused by:
        org.mockito.exceptions.base.MockitoException: Could not modify all classes [class java.lang.Object, class batect.logging.Logger]
            at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:137)
            at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:344)
            at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:159)
            at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:353)
            ... 7 more

            Caused by:
            java.lang.IllegalStateException: 
            Byte Buddy could not instrument all classes within the mock's type hierarchy

            This problem should never occur for javac-compiled classes. This problem has been observed for classes that are:
             - Compiled by older versions of scalac
             - Classes that are part of the Android distribution
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:122)
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.mockClass(InlineBytecodeGenerator.java:99)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:37)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:34)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:137)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:344)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:159)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:353)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:32)
                at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMockType(InlineByteBuddyMockMaker.java:200)
                at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMock(InlineByteBuddyMockMaker.java:181)
                at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:35)
                at org.mockito.internal.MockitoCore.mock(MockitoCore.java:65)
                at org.mockito.Mockito.mock(Mockito.java:1875)
                ... 7 more

                Caused by:
                java.lang.IllegalArgumentException: Unsupported class file major version 55
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:166)
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:148)
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:136)
                    at net.bytebuddy.utility.OpenedClassReader.of(OpenedClassReader.java:54)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:2912)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:1634)
                    at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:171)
                    at net.bytebuddy.dynamic.scaffold.inline.AbstractInliningDynamicTypeBuilder.make(AbstractInliningDynamicTypeBuilder.java:92)
                    at net.bytebuddy.dynamic.DynamicType$Builder$AbstractBase.make(DynamicType.java:2669)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.transform(InlineBytecodeGenerator.java:183)
                    at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246)
                    at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
                    at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:563)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:167)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:119)
                    ... 20 more

1 test completed, 1 failed
```

**Comment by charleskorn** [comment]
Hmmm... if you run the unit tests (`./gradlew check`), do they pass, or do they fail for the same reason?

**Comment by binkley** [comment]
Howdy.  Int tests fail with:
```
batect.integrationtests.DockerClientIntegrationTest > describe a Docker client FAILED
    org.mockito.exceptions.base.MockitoException: 
    Mockito cannot mock this class: class batect.logging.Logger.
    Can not mock final classes with the following settings :
     - explicit serialization (e.g. withSettings().serializable())
     - extra interfaces (e.g. withSettings().extraInterfaces(...))

    You are seeing this disclaimer because Mockito is configured to create inlined mocks.
    You can learn about inline mocks and their limitations under item #39 of the Mockito class javadoc.

    Underlying exception : org.mockito.exceptions.base.MockitoException: Could not modify all classes [class java.lang.Object, class batect.logging.Logger]
        at batect.integrationtests.DockerClientIntegrationTest$1$1.invoke(DockerClientIntegrationTest.kt:334)
        at batect.integrationtests.DockerClientIntegrationTest$1$1.invoke(DockerClientIntegrationTest.kt:68)
        at org.jetbrains.spek.engine.SpekTestEngine$Collector.group(SpekTestEngine.kt:152)
        at org.jetbrains.spek.api.dsl.SpecBody$DefaultImpls.group$default(SpecBody.kt:13)
        at org.jetbrains.spek.api.dsl.StandardKt.describe(Standard.kt:10)
        at batect.integrationtests.DockerClientIntegrationTest$1.invoke(DockerClientIntegrationTest.kt:69)
        at batect.integrationtests.DockerClientIntegrationTest$1.invoke(DockerClientIntegrationTest.kt:68)

        Caused by:
        org.mockito.exceptions.base.MockitoException: Could not modify all classes [class java.lang.Object, class batect.logging.Logger]
            at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:137)
            at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:344)
            at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:159)
            at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:353)
            ... 7 more

            Caused by:
            java.lang.IllegalStateException: 
            Byte Buddy could not instrument all classes within the mock's type hierarchy

            This problem should never occur for javac-compiled classes. This problem has been observed for classes that are:
             - Compiled by older versions of scalac
             - Classes that are part of the Android distribution
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:122)
                at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.mockClass(InlineBytecodeGenerator.java:99)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:37)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:34)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:137)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:344)
                at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:159)
                at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:353)
                at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:32)
                at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMockType(InlineByteBuddyMockMaker.java:200)
                at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMock(InlineByteBuddyMockMaker.java:181)
                at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:35)
                at org.mockito.internal.MockitoCore.mock(MockitoCore.java:65)
                at org.mockito.Mockito.mock(Mockito.java:1875)
                ... 7 more

                Caused by:
                java.lang.IllegalArgumentException: Unsupported class file major version 55
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:166)
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:148)
                    at net.bytebuddy.jar.asm.ClassReader.<init>(ClassReader.java:136)
                    at net.bytebuddy.utility.OpenedClassReader.of(OpenedClassReader.java:54)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:2912)
                    at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:1634)
                    at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:171)
                    at net.bytebuddy.dynamic.scaffold.inline.AbstractInliningDynamicTypeBuilder.make(AbstractInliningDynamicTypeBuilder.java:92)
                    at net.bytebuddy.dynamic.DynamicType$Builder$AbstractBase.make(DynamicType.java:2669)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.transform(InlineBytecodeGenerator.java:183)
                    at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246)
                    at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
                    at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:563)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
                    at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:167)
                    at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:119)
                    ... 20 more
```

**Comment by charleskorn** [comment]
That's very odd. If it was a Java 11 issue, then the unit tests wouldn't pass either, but you're getting to the integration tests so the unit tests are passing. Can you verify that [this file](https://github.com/charleskorn/batect/blob/master/app/src/integrationTest/resources/mockito-extensions/org.mockito.plugins.MockMaker) is present in your clone and has the same content?

**Comment by charleskorn** [comment]
Can you try again with the latest master (https://github.com/charleskorn/batect/commit/91f7ab946889a2ad8069aab0b18380215d6939e7)? 

I've upgraded to a newer version of mockito-kotlin (which in turn brings in a newer version of Mockito) that should have some Java 11-related fixes.

**Comment by binkley** [comment]
Cool beans.  Int tests pass; now I'm watching shellcheck do it's thing.  Thanks!

**Comment by binkley** [comment]
This looks a little more like Java version (change in class file format number):
```
> Task :wrapper:test
.F..F.
======================================================================
FAIL: test_first_run (__main__.WrapperScriptTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 30, in test_first_run
    self.assertIn("BATECT_WRAPPER_SCRIPT_PATH is: {}\n".format(self.get_script_path()), output)
AssertionError: 'BATECT_WRAPPER_SCRIPT_PATH is: /Users/boxley/src/kt/batect/wrapper/src/template.sh\n' not found in 'Downloading batect version VERSION-GOES-HERE from http://localhost:8080/test/testapp.jar...\n\r######################################################################## 100.0%\nError: LinkageError occurred while loading main class testapp.Application\n\tjava.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 54.0\n'

======================================================================
FAIL: test_second_run (__main__.WrapperScriptTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 39, in test_second_run
    self.assertIn("BATECT_WRAPPER_SCRIPT_PATH is: {}\n".format(self.get_script_path()), first_output)
AssertionError: 'BATECT_WRAPPER_SCRIPT_PATH is: /Users/boxley/src/kt/batect/wrapper/src/template.sh\n' not found in 'Downloading batect version VERSION-GOES-HERE from http://localhost:8080/test/testapp.jar...\n\r######################################################################## 100.0%\nError: LinkageError occurred while loading main class testapp.Application\n\tjava.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 54.0\n'

======================================================================
FAIL: test_supported_java (__main__.WrapperScriptTests) (java_version='8')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 92, in test_supported_java
    self.assertIn("The Java application has started.", result.stdout.decode())
AssertionError: 'The Java application has started.' not found in 'Downloading batect version VERSION-GOES-HERE from http://localhost:8080/test/testapp.jar...\n\r######################################################################## 100.0%\nError: A JNI error has occurred, please check your installation and try again\nException in thread "main" java.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0\n\tat java.lang.ClassLoader.defineClass1(Native Method)\n\tat java.lang.ClassLoader.defineClass(ClassLoader.java:763)\n\tat java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)\n\tat java.net.URLClassLoader.defineClass(URLClassLoader.java:468)\n\tat java.net.URLClassLoader.access$100(URLClassLoader.java:74)\n\tat java.net.URLClassLoader$1.run(URLClassLoader.java:369)\n\tat java.net.URLClassLoader$1.run(URLClassLoader.java:363)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat java.net.URLClassLoader.findClass(URLClassLoader.java:362)\n\tat java.lang.ClassLoader.loadClass(ClassLoader.java:424)\n\tat sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)\n\tat java.lang.ClassLoader.loadClass(ClassLoader.java:357)\n\tat sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:495)\n'

======================================================================
FAIL: test_supported_java (__main__.WrapperScriptTests) (java_version='9')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 92, in test_supported_java
    self.assertIn("The Java application has started.", result.stdout.decode())
AssertionError: 'The Java application has started.' not found in 'Error: LinkageError occurred while loading main class testapp.Application\n\tjava.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 53.0\n'

======================================================================
FAIL: test_supported_java (__main__.WrapperScriptTests) (java_version='10')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 92, in test_supported_java
    self.assertIn("The Java application has started.", result.stdout.decode())
AssertionError: 'The Java application has started.' not found in 'Error: LinkageError occurred while loading main class testapp.Application\n\tjava.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 54.0\n'

======================================================================
FAIL: test_supported_java (__main__.WrapperScriptTests) (java_version='11')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/tests.py", line 92, in test_supported_java
    self.assertIn("The Java application has started.", result.stdout.decode())
AssertionError: 'The Java application has started.' not found in 'Error: LinkageError occurred while loading main class testapp.Application\n\tjava.lang.UnsupportedClassVersionError: testapp/Application has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 54.0\n'
```

**Comment by charleskorn** [comment]
I've just pushed a commit that should fix this (https://github.com/charleskorn/batect/commit/e42f66783996f27d00c2345838d840674a75d962 - I referenced the wrong issue in the commit message) - the target Java version was set for the main application but not the test app used for the wrapper.

**Comment by binkley** [comment]
I'll try again, thank you, and report back.

**Comment by charleskorn** [comment]
No worries, let me know how you go.

**Comment by binkley** [comment]
Pulled from upstream into my fork.  Ran `./gradlew clean check`.  I do have a test failure, but it no longer looks related to JVM 11.  I'm guessing I need to poke at my Docker network settings (I using Docker Desktop 2.0.0.0-mac81, stable channel):
```
batect.integrationtests.DockerClientIntegrationTest > on the image having an EXPOSE instruction for the port to be exposed FAILED
    java.io.IOException: unexpected end of stream on Connection{localhost:8080, proxy=DIRECT hostAddress=localhost/0:0:0:0:0:0:0:1:8080 cipherSuite=none protocol=http/1.1}
        at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:208)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
        at okhttp3.RealCall.execute(RealCall.java:92)
        at batect.integrationtests.DockerClientIntegrationTestKt.httpGet(DockerClientIntegrationTest.kt:319)
        at batect.integrationtests.DockerClientIntegrationTestKt.access$httpGet(DockerClientIntegrationTest.kt:1)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1$response$1$1.invoke(DockerClientIntegrationTest.kt:244)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1$response$1$1.invoke(DockerClientIntegrationTest.kt:69)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$6.invoke(DockerClientIntegrationTest.kt:145)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1$response$1.invoke(DockerClientIntegrationTest.kt:241)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1$response$1.invoke(DockerClientIntegrationTest.kt:69)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$5.invoke(DockerClientIntegrationTest.kt:135)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1.invoke(DockerClientIntegrationTest.kt:240)
        at batect.integrationtests.DockerClientIntegrationTest$1$1$12$1.invoke(DockerClientIntegrationTest.kt:69)

        Caused by:
        java.io.EOFException: \n not found: limit=0 content=…
            at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:236)
            at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215)
            at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
            ... 25 more

10 tests completed, 1 failed
```

**Comment by charleskorn** [comment]
Unfortunately that's a manifestation of https://github.com/square/okhttp/issues/2738 that comes up from time to time. It seems to only affect OS X (the CI build that runs on Linux never has the same problem). It's been on my list to investigate further but happens infrequently enough that it hasn't been a priority. Re-running the tests usually fixes the problem (which I know isn't really a fix :) )

**Comment by charleskorn** [comment]
I've added a fix for this in d6a1ebb - please pull the latest master and let me know if you still run into this issue.

If everything's working, please go ahead and close this issue, otherwise, please post back here with any further problems.

Thanks again for your feedback!

**Comment by charleskorn** [comment]
I'm going to go ahead and close this due to inactivity - if you run into it again, please reopen this issue.
